File: libs/dotnet/SquareSums/QSortNodesSorting.cs
using System;

// ReSharper disable ForCanBeConvertedToForeach

namespace SquareSums
{
    public class QSortNodesSorting : INodesSorting
    {
        private readonly NodesComparer _comparer;

        public QSortNodesSorting(Path? path, int maxN)
        {
            _comparer = new NodesComparer(path, maxN);
        }

        public void SortNodes(Span<Node> nodes)
        {
            _comparer.FlushCache();

            Quicksort(_comparer, nodes, 0, nodes.Length - 1);
        }

        private static int Partition(NodesComparer comparer, Span<Node> nodes, int start, int end)
        {
            int marker = start;
            for (int i = start; i < end; i++)
            {
                var compareResult = comparer.Compare(nodes[i], nodes[end]);
                if (compareResult == -1)
                {
                    (nodes[marker], nodes[i]) = (nodes[i], nodes[marker]);
                    marker += 1;
                }
            }
            
            (nodes[marker], nodes[end]) = (nodes[end], nodes[marker]);
            return marker;
        }

        static void Quicksort(NodesComparer comparer, Span<Node> nodes, int start, int end)
        {
            if (start >= end)
            {
                return;
            }

            int pivot = Partition(comparer, nodes, start, end);
            Quicksort(comparer, nodes, start, pivot - 1);
            Quicksort(comparer, nodes, pivot + 1, end);
        }
    }
}

File: libs/js/src/QSortNodesSorting.ts
import { Path } from './Path';
import { Node } from './Node';
import { NodesComparer } from './NodesComparer';

function partition(comparer: NodesComparer, nodes: Array<Node>, start: number, end: number): number {
    let marker = start;
    for (let i = start; i < end; i++) {
        var compareResult = comparer.compare(nodes[i], nodes[end]);
        if (compareResult == -1) {
            let temp = nodes[marker]; 
            nodes[marker] = nodes[i];
            nodes[i] = temp;
            marker += 1;
        }
    }

    let temp = nodes[marker];
    nodes[marker] = nodes[end];
    nodes[end] = temp; 
    return marker;
}

function quicksort(comparer: NodesComparer, nodes: Array<Node>, start: number, end: number) {
    if (start >= end) {
        return;
    }

    let pivot = partition(comparer, nodes, start, end);
    quicksort(comparer, nodes, start, pivot - 1);
    quicksort(comparer, nodes, pivot + 1, end);
}

export class QSortNodesSorting {
    private _comparer: NodesComparer;

    constructor(path: Path | null, maxN: number) {
        this._comparer = new NodesComparer(path, maxN);
    }

    public sortNodes(nodes: Array<Node>) {
        this._comparer.flushCache();
        quicksort(this._comparer, nodes, 0, nodes.length - 1);
    }
}